src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -68,187 +66,192 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id_t pg_owner,
+        update_shard_in_map(shard_info);
         break;
     }
     default: {
         break;
     }
     }
 
-    if (ctx) { ctx->promise_.setValue(ShardManager::Result< ShardInfo >(shard.info)); }
+    if (ctx) { ctx->promise_.setValue(ShardManager::Result< ShardInfo >(shard_info)); }
 }
 
-void HSHomeObject::do_commit_new_shard(const Shard& shard) {
+void HSHomeObject::add_new_shard_to_map(ShardPtr&& shard) {
     // TODO: We are taking a global lock for all pgs to create shard. Is it really needed??
     // We need to have fine grained per PG lock and take only that.
     std::scoped_lock lock_guard(_pg_lock, _shard_lock);
-    auto pg_iter = _pg_map.find(shard.info.placement_group);
+    auto pg_iter = _pg_map.find(shard->info.placement_group);
     RELEASE_ASSERT(pg_iter != _pg_map.end(), "Missing PG info");
     auto& shards = pg_iter->second->shards_;
-    auto iter = shards.emplace(shards.end(), shard);
-    auto [_, happened] = _shard_map.emplace(shard.info.id, iter);
+    auto shard_id = shard->info.id;
+    auto iter = shards.emplace(shards.end(), std::move(shard));
+    auto [_, happened] = _shard_map.emplace(shard_id, iter);
     RELEASE_ASSERT(happened, "duplicated shard info");
 
-    // following part is must for follower members or when member is restarted;
-    auto sequence_num = get_sequence_num_from_shard_id(shard.info.id);
+    // following part gives follower members a chance to catch up shard sequence num;
+    auto sequence_num = get_sequence_num_from_shard_id(shard_id);
     if (sequence_num > pg_iter->second->shard_sequence_num_) { pg_iter->second->shard_sequence_num_ = sequence_num; }
 }
 
-void HSHomeObject::do_commit_seal_shard(const Shard& shard) {
+void HSHomeObject::update_shard_in_map(const ShardInfo& shard_info) {
     std::scoped_lock lock_guard(_shard_lock);
-    auto shard_iter = _shard_map.find(shard.info.id);
+    auto shard_iter = _shard_map.find(shard_info.id);
     RELEASE_ASSERT(shard_iter != _shard_map.end(), "Missing shard info");
-    *(shard_iter->second) = shard;
+    auto hs_shard = d_cast< HS_Shard* >((*shard_iter->second).get());
+    hs_shard->update_info(shard_info);
 }
 
-void* HSHomeObject::get_shard_metablk(shard_id_t id) {
+ShardManager::Result< homestore::chunk_num_t > HSHomeObject::get_shard_chunk(shard_id_t id) const {
     std::scoped_lock lock_guard(_shard_lock);
     auto shard_iter = _shard_map.find(id);
-    if (shard_iter == _shard_map.end()) { return nullptr; }
-    return (*shard_iter->second).metablk_cookie;
+    if (shard_iter == _shard_map.end()) { return folly::makeUnexpected(ShardError::UNKNOWN_SHARD); }
+    auto hs_shard = d_cast< HS_Shard* >((*shard_iter->second).get());
+    return hs_shard->sb_->chunk_id;
+}
+
+std::optional< homestore::chunk_num_t > HSHomeObject::get_any_chunk_id(pg_id_t const pg_id) {
+    std::scoped_lock lock_guard(_pg_lock);
+    auto pg_iter = _pg_map.find(pg_id);
+    RELEASE_ASSERT(pg_iter != _pg_map.end(), "Missing PG info");
```

Comment:
Suggest not do assert, but return empty and let caller to check if this pg_id has been created in the map before, since ret value is `std::optional`, which also gives caller some flexibility to call with pg_id that may or may not be added to the map yet.

Reply:
such cases are added now.